### PR TITLE
Manage a nic used for north/south traffic

### PR DESCRIPTION
**Pull request by gnuoy**
If a nic is supplied for use for north/south traffic then wire it into br-ex. Also manage removing it if /reset is called.

### CODE REVIEW

**Comment by gnuoy** [comment]
Corresponding microstack PR https://github.com/openstack-snaps/snap-microstack/pull/55

File: openstack_hypervisor/api.py [hunk 1]
```diff
@@ -141,6 +142,14 @@ async def update_logging(config: model.LoggingConfig):
...
 async def reset_config():
     """Reset all configs to default."""
+    try:
+        # Need to remove the nic from the bridge before the config is reset and the
+        # nic name is lost.
+        gateway_nic = snap.config.get("network.gateway-nic")
+        external_bridge = snap.config.get("network.external-bridge")
+        hooks.del_interface_from_bridge(external_bridge, gateway_nic)
+    except UnknownConfigKey:
+        pass
```
Comment (javacruft) [inline openstack_hypervisor/api.py]: will and UNSET key throw a UnknownConfigKey error?  if not then we should probably validate that the nic option is actually set

**Comment by javacruft** [comment]
One inline comment and a general observation; we have the opportunity to record which interface was configured by the snap in OVS as part of the additional metadata we can store in OVS.

This would allow us to detect if the configuration option was changed - the old interface can be removed and the new one added thus avoiding someone creating a spanning tree loop in their network or other such horribleness.

This is implemented in the ovn charm layer for the existing machine charms:

https://github.com/openstack-charmers/charm-layer-ovn/blob/master/lib/charms/ovn_charm.py#L1246

File: openstack_hypervisor/hooks.py [hunk 2]
```diff
@@ -499,6 +500,105 @@ def _configure_ovn_base(snap: Snap) -> None:
...
+    """
+    external_ports = _get_external_ports_on_bridge(external_bridge)
+    if external_nic in external_ports:
+        logging.debug(f"{external_nic} already attached to {external_bridge}")
+    else:
+        _add_interface_to_bridge(external_bridge, external_nic)
+    for p in external_ports:
+        if p != external_nic:
+            logging.debug(
+                f"Removing additional extranal ports {external_ports} from {external_bridge}"
```
Comment (javacruft) [inline openstack_hypervisor/hooks.py]: typo: extranal -> external

Comment (javacruft) [reply openstack_hypervisor/hooks.py]: and should this be {p} rather than the full list?

**Review by javacruft — APPROVED** [review]

**Pull request merged**